pkg/distributor/distributor.go

```diff
@@ -908,18 +893,41 @@ func (d *Distributor) PushWithCleanup(ctx context.Context, req *mimirpb.WriteReq
 		return d.send(localCtx, ingester, timeseries, metadata, req.Source)
 	}, func() { cleanup(); cancel() })
 
-	if forwardingErrCh != nil {
-		// Blocks until the forwarding requests have completed and the final status has been pushed through this chan.
-		forwardingErr := <-forwardingErrCh
-		err = httpgrpcutil.PrioritizeRecoverableErr(err, forwardingErr, firstPartialErr)
-	}
-
 	if err != nil {
 		return nil, err
 	}
 	return &mimirpb.WriteResponse{}, firstPartialErr
 }
 
+// forwardingReq returns a forwarding request if one is necessary, given the user ID.
+// if no forwarding request is necessary it returns nil.
+func (d *Distributor) forwardSamples(ctx context.Context, req *mimirpb.WriteRequest, userID string, cleanup func()) (*mimirpb.WriteResponse, error) {
+	forwardingRules := d.limits.ForwardingRules(userID)
```

Comment:
yeah, there's currently a bug because I only hooked it up with the `Push` method but actually the API uses `PushWithCleanup` ([Oleg's comment](https://github.com/grafana/mimir/pull/2295/#issuecomment-1195556150)). I'll fix that.

Reply:
I changed this part completely.

I copied a lot of the ideas from Dimitar's [PR](https://github.com/grafana/mimir/pull/2419), basically the `PushWithCleanup` method now gets wrapped with an optional wrapper in the API, this should also lead to a more consistent solution once this PR and Dimitar's PR are merged: 

https://github.com/grafana/mimir/blob/fdfab1479188d4843299e3ff48c910906029c91c/pkg/api/api.go#L239-L245